services/udp/udp.go

```diff
@@ -0,0 +1,147 @@
+package udp
+
+import (
+	"errors"
+	"log"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/influxdb/influxdb/cluster"
+	"github.com/influxdb/influxdb/tsdb"
+)
+
+const (
+	UDPBufferSize = 65536
+)
+
+type Server struct {
+	conn *net.UDPConn
+	addr *net.UDPAddr
+	wg   sync.WaitGroup
+	done chan struct{}
+
+	batcher *tsdb.PointBatcher
+
+	batchSize    int
+	batchTimeout time.Duration
+	database     string
+
+	PointsWriter interface {
+		WritePoints(p *cluster.WritePointsRequest) error
+	}
+
+	Logger *log.Logger
+}
+
+func NewServer(db string) *Server {
+	return &Server{
+		done:     make(chan struct{}),
+		database: db,
+		Logger:   log.New(os.Stderr, "[udp] ", log.LstdFlags),
+	}
+}
+
+func (s *Server) SetBatchSize(sz int)             { s.batchSize = sz }
+func (s *Server) SetBatchTimeout(d time.Duration) { s.batchTimeout = d }
+
+func (s *Server) ListenAndServe(iface string) (err error) {
+	if iface == "" {
+		return errors.New("bind address has to be specified in config")
+	}
+	if s.database == "" {
+		return errors.New("database has to be specified in config")
+	}
+
+	s.addr, err = net.ResolveUDPAddr("udp", iface)
+	if err != nil {
+		s.Logger.Printf("Failed to resolve UDP address %s: %s", iface, err)
+		return err
+	}
+
+	s.conn, err = net.ListenUDP("udp", s.addr)
+	if err != nil {
+		s.Logger.Printf("Failed to set up UDP listener at address %s: %s", s.addr, err)
+		return err
+	}
+
+	s.Logger.Printf("Started listening on %s", iface)
+
+	s.batcher = tsdb.NewPointBatcher(s.batchSize, s.batchTimeout)
+
+	s.wg.Add(2)
+	go s.serve()
+	go s.writePoints()
+
+	return nil
+}
+
+func (s *Server) writePoints() {
+	defer s.wg.Done()
+
+	for {
+		select {
+		case batch := <-s.batcher.Out():
+			err := s.PointsWriter.WritePoints(&cluster.WritePointsRequest{
+				Database:         s.database,
+				RetentionPolicy:  "",
+				ConsistencyLevel: cluster.ConsistencyLevelOne,
+				Points:           batch,
+			})
+			if err != nil {
+				s.Logger.Printf("Failed to write point batch to database %q: %s\n", s.database, err)
+			} else {
+				s.Logger.Printf("Wrote a batch of %d points to %s", len(batch), s.database)
+			}
+		case <-s.done:
+			return
+		}
+	}
+}
+
+func (s *Server) serve() {
+	defer s.wg.Done()
+
+	s.batcher.Start()
+	for {
+		buf := make([]byte, UDPBufferSize)
+		n, _, err := s.conn.ReadFromUDP(buf)
+		if err != nil {
+			s.Logger.Printf("Failed to read UDP message: %s", err)
+			s.batcher.Flush()
+			return
```

Comment:
@corylanou Should I continue instead (just like in collectd) ? And what about the call to batcher.Flush() ?